drivers/input/keyboard_upper.c

```diff
@@ -0,0 +1,435 @@
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static int     keyboard_open(FAR struct file *filep);
+static int     keyboard_close(FAR struct file *filep);
+static ssize_t keyboard_read(FAR struct file *filep, FAR char *buffer,
+                             size_t len);
+static ssize_t keyboard_write(FAR struct file *filep, FAR const char *buffer,
+                              size_t len);
+static int     keyboard_poll(FAR struct file *filep, FAR struct pollfd *fds,
+                             bool setup);
+static void    keyboard_notify(FAR struct keyboard_opriv_s *buffer);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static const struct file_operations g_keyboard_fops =
+{
+  keyboard_open,  /* open */
+  keyboard_close, /* close */
+  keyboard_read,  /* read */
+  keyboard_write, /* write */
+  NULL,           /* seek */
+  NULL,           /* ioctl */
+  keyboard_poll   /* poll */
+#ifndef CONFIG_DISABLE_PSEUDOFS_OPERATIONS
+  , NULL  /* unlink */
+#endif
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: keyboard_notify
+ ****************************************************************************/
+
+static void keyboard_notify(FAR struct keyboard_opriv_s *opriv)
+{
+  FAR struct pollfd *fds = opriv->fds;
+  if (fds != NULL)
+    {
+      fds->revents |= (fds->events & POLLIN);
+      if (fds->revents != 0)
+        {
+          /* report event log */
+
+          nxsem_post(fds->sem);
+        }
+    }
+}
+
+/****************************************************************************
+ * Name: keyboard_open
+ ****************************************************************************/
+
+static int keyboard_open(FAR struct file *filep)
+{
+  FAR struct keyboard_opriv_s     *opriv = NULL;
+  FAR struct inode                *inode = filep->f_inode;
+  FAR struct keyboard_upperhalf_s *upper = inode->i_private;
+  int ret;
+
+  ret = nxsem_wait(&upper->exclsem);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  /* Perform the operations required by the lower level */
+
+  if (upper->lower->open)
+    {
+      ret = upper->lower->open(upper->lower);
+      if (ret < 0)
+        {
+          goto out;
+        }
+    }
+
+  opriv = kmm_zalloc(sizeof(FAR struct keyboard_opriv_s));
+  if (opriv == NULL)
+    {
+      ret = -ENOMEM;
+      goto out;
+    }
+
+  /* Initializes the buffer for each open file */
+
+  ret = circbuf_init(&opriv->circ, NULL, CONFIG_INPUT_KEYBOARD_BUFFSIZE);
+  if (ret < 0)
+    {
+      kmm_free(opriv);
+      goto out;
+    }
```

Comment:
Can we reorder this so `upper->lower->open(upper->lower);` will only be called if all allocations are successful?

Reply:
done